src/api/db/models/User.ts

```diff
@@ -0,0 +1,277 @@
+    const cognito = new Cognito.CognitoIdentityProviderClient();
+
+    try {
+      const users = (
+        await Promise.all(
+          ['manager', 'observer', 'member'].map(async (role) => {
+            const list = [];
+            let res: Cognito.ListUsersInGroupCommandOutput;
+            do {
+              res = await cognito.send(
+                new Cognito.ListUsersInGroupCommand({
+                  Limit: 60,
+                  UserPoolId: context.config['/APPLICATION/COGNITO/USERPOOLID'],
+                  GroupName: `animl/${context.user['curr_project']}/project_${role}`,
+                }),
+              );
+
+              list.push(...(res.Users || []));
+            } while (res.NextToken);
+
+            return list.map((user): UserWithRole => {
+              (user as UserWithRole).role = role;
+              return user as UserWithRole;
+            });
+          }),
+        )
+      )
+        .reduce((acc, cur) => acc.concat(cur), [])
+        .map((user) => ({
+          role: user.role,
+          username: user.Username!,
+          email: user.Attributes?.find(({ Name }) => Name === 'email')?.Value,
+          created: user.UserCreateDate,
+          updated: user.UserLastModifiedDate,
+          enabled: user.Enabled,
+          status: user.UserStatus,
+        }))
+        .filter((user) => !input.filter || user.username.includes(input.filter));
+
+      const roles = new Map<string, CognitoUser>();
+      for (const { role, username, ...user } of users) {
+        if (roles.has(username)) {
+          roles.get(username)!.roles.push(role);
+        } else {
+          roles.set(username, { ...user, username, roles: [role] });
+        }
+      }
+
+      return {
+        users: Array.from(roles.values()),
+      };
+    } catch (err) {
+      if (err instanceof GraphQLError) throw err;
+      throw new InternalServerError(err as string);
+    }
+  }
+}
+
+export default class AuthedUserModel extends BaseAuthedModel {
+  @roleCheck(MANAGE_USERS_ROLES)
+  createGroups(...args: MethodParams<typeof UserModel.createGroups>) {
+    return UserModel.createGroups(...args);
+  }
+
+  @roleCheck(MANAGE_USERS_ROLES)
+  createUser(...args: MethodParams<typeof UserModel.create>) {
+    return UserModel.create(...args);
+  }
+
+  @roleCheck(MANAGE_USERS_ROLES)
+  listUsers(...args: MethodParams<typeof UserModel.list>) {
+    return UserModel.list(...args);
+  }
+
+  @roleCheck(MANAGE_USERS_ROLES)
+  updateUser(...args: MethodParams<typeof UserModel.update>) {
+    return UserModel.update(...args);
+  }
+}
+
+type UserWithRole = Cognito.UserType & { role: string };
+interface CognitoUser {
+  roles: string[];
+  username: string;
+  email: string | undefined;
+  created: Date | undefined;
+  updated: Date | undefined;
+  enabled: boolean | undefined;
+  status: Cognito.UserStatusType | undefined;
+}
+
+interface CreateGroupsInput {
+  name: string;
+}
+
+interface CreateUserInput {
```

Comment:
@alukach aren't some of these input types already being generated by the GraphQL codegen?

Reply:
Yeah, pretty much the same thing as https://github.com/tnc-ca-geo/animl-api/pull/208#discussion_r1640172109.

In [ecf7fb1](https://github.com/tnc-ca-geo/animl-api/pull/207/commits/ecf7fb1905aef16a3062ead26b353987ca7a9a8e), I've reworked this to utilize the generated GQL types.